### PR TITLE
[BugFix] Preserve batch invariance for mixed prefill/decode and doc clarify

### PR DESCRIPTION
**Pull request by wangx700**
### What this PR does / why we need it?

ref to https://github.com/vllm-project/vllm-ascend/issues/10072

This PR fixes batch-invariance gaps in mixed prefill/decode attention and NPU sum reductions:

- Keeps the original inline A5 ChunkedPrefill mixed prefill/decode split condition and adds `VLLM_BATCH_INVARIANT` as an alternative control. When batch invariance is enabled, mixed prefill/decode batches use separate FIA calls regardless of device generation or attention state; when it is disabled, the existing A5 behavior is unchanged for performance optimization.
- Keeps the existing `reduce_sum` and `torch.sum` batch-invariant behavior unchanged.
- Reuses the existing `reduce_sum` implementation directly for both `torch.sum` and `Tensor.sum`, without an extra forwarding adapter.
- The existing `topk_weights.sum(dim=-1, keepdim=True)` call sites remain unchanged and automatically follow `Tensor.sum -> reduce_sum -> npu_reduce_sum_batch_invariant`.
- The batch-invariance guide documents the two-stage `batch_invariant_ops` installation flow: first build and install the AscendC operator run package, then build and install the wheel that exposes the operator through PyTorch.

### Documentation changes

- Updated the installation guidance to clarify that `VLLM_BATCH_INVARIANT=1` is consumed by source builds (`pip install .` and `pip install -e .`) with `COMPILE_CUSTOM_KERNELS=1`; installing a prebuilt wheel does not rebuild the operators.
- Expanded `docs/source/user_guide/feature_guide/batch_invariance.md` with separate source-build and post-install operator setup instructions. The guide clarifies that prebuilt wheels do not contain `csrc/build_batch_invariant_ops.sh`, so post-install setup requires a matching source checkout. It also documents that A2/A3 Docker images install the operator packages at image-build time while runtime activation still requires `VLLM_BATCH_INVARIANT=1`.
- Documented the ordered two-stage operator build flow:
  1. Build and install the operator run package that provides the device-side AscendC implementation.
  2. Build and install the `batch_invariant_ops` wheel that exposes the AscendC operators through PyTorch.
- Added the A2 (`ascend910b`) and A3 (`ascend910_93`) build commands and a quick import/operator availability check after installation.

### Does this PR introduce _any_ user-facing change?

Yes. With `VLLM_BATCH_INVARIANT=1`:

- Mixed prefill/decode attention is processed in separate prefill and decode FIA calls.
- `Tensor.sum(dim=-1, keepdim=True)` follows the existing batch-invariant `reduce_sum` path without modifying its business call sites.

### How was this patch tested?
- A5 operator profiling collection verification: after enabling VLLM_BATCH_INVARIANT=1, topk_weights.sum(dim=-1, keepdim=True) is replaced with npu_reduce_sum_batch_invariant.
- Added coverage confirming that AscendC mode binds both `torch.sum` and `Tensor.sum` directly to `reduce_sum`; existing NPU and CPU tests cover the operator and native fallback paths.
- Added unit coverage confirming that NPU `reduce_sum` calls `npu_reduce_sum_batch_invariant` and CPU `reduce_sum` calls the saved native `torch.sum` implementation.
```
root@C04-B03-OS1:/mnt/share/w00899129/vllm_code/vllm-ascend# python  tests/ut/test_batch_invariant.py
INFO 08-31 15:24:36 [__init__.py:52] Available plugins for group vllm.platform_plugins:
INFO 08-31 15:24:36 [__init__.py:54] - ascend -> vllm_ascend:register
INFO 08-31 15:24:36 [__init__.py:57] All plugins in this group will be loaded. Set `VLLM_PLUGINS` to control which plugins to load.
INFO 08-31 15:24:36 [__init__.py:254] Platform plugin ascend is activated
INFO 08-31 15:24:36 [platform.py:70] Breakable CUDAGraph on Ascend is opt-in; using VLLM_USE_BREAKABLE_CUDAGRAPH=0.
================================================================== test session starts ===================================================================
platform linux -- Python 3.11.10, pytest-8.3.2, pluggy-1.6.0
rootdir: /mnt/share/w00899129/vllm_code/vllm-ascend
configfile: pyproject.toml
plugins: xdist-3.6.1, anyio-4.14.2
collected 13 items                                                                                                                                       

tests/ut/test_batch_invariant.py .............                                                                                                     [100%]

==================================================================== warnings summary ====================================================================
../../../../../usr/local/python3.11.10/lib/python3.11/site-packages/_pytest/config/__init__.py:1273
  /usr/local/python3.11.10/lib/python3.11/site-packages/_pytest/config/__init__.py:1273: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: anyio
    self._mark_plugins_for_rewrite(hook)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================= 13 passed, 1 warning in 0.29s ==============================================================
root@C04-B03-OS1:/mnt/share/w00899129/vllm_code/vllm-ascend# 
```

- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/batch_invariant.py [hunk 1]
```diff
@@ -73,6 +73,11 @@ def reduce_sum(x: torch.Tensor, dim: int | None = None, keepdim: bool = False) -
     return torch_sum(x, dim, keepdim)
 
 
+def tensor_sum(x: torch.Tensor, dim: int | None = None, keepdim: bool = False) -> torch.Tensor:
```
Comment (lcfenglinwan) [inline vllm_ascend/batch_invariant.py]: check if it is necessary

**Review by linfeng-yuan — DISMISSED** [review]

File: vllm_ascend/batch_invariant.py (hunk 1 above)
Comment (wangx700) [reply vllm_ascend/batch_invariant.py]: OK, it is not necessary after tested.

File: docs/source/user_guide/feature_guide/batch_invariance.md [hunk 2]
```diff
@@ -28,7 +28,59 @@ We will support Ascend 950 Products and other NPUs in the future.
...
+
+The `batch_invariant_ops` wheel are built into the vllm-ascend installation flow. Set `VLLM_BATCH_INVARIANT=1` to enable the building of `batch_invariant_ops` wheel:
+
+```bash
+# original
+pip install \
+    --extra-index-url https://mirrors.huaweicloud.com/ascend/repos/pypi/variant \
+    vllm-ascend=={{ pip_vllm_ascend_version }}
+# set `VLLM_BATCH_INVARIANT=1` before building vllm-ascend to build `batch_invariant_ops` wheel.
+VLLM_BATCH_INVARIANT=1 pip install \
```
Comment (Ronald1995) [inline docs/source/user_guide/feature_guide/batch_invariance.md]: please make sure we the var is valid in pip install and pip install -e

File: docs/source/user_guide/feature_guide/batch_invariance.md [hunk 3]
```diff
@@ -28,7 +28,59 @@ We will support Ascend 950 Products and other NPUs in the future.
...
+cd <vllm-ascend-install-dir>
+bash csrc/build_batch_invariant_ops.sh ascend910b
+```
+
+**A3:**
+
+```bash
+cd <vllm-ascend-install-dir>
+bash csrc/build_batch_invariant_ops.sh ascend910_93
+```
```
Comment (Ronald1995) [inline docs/source/user_guide/feature_guide/batch_invariance.md:76]: please make sure it will suit for `pip install` and `pip install -e`

File: docs/source/user_guide/feature_guide/batch_invariance.md [hunk 4]
```diff
@@ -28,7 +28,59 @@ We will support Ascend 950 Products and other NPUs in the future.
...
+pip install \
+    --extra-index-url https://mirrors.huaweicloud.com/ascend/repos/pypi/variant \
+    vllm-ascend=={{ pip_vllm_ascend_version }}
+# set `VLLM_BATCH_INVARIANT=1` before building vllm-ascend to build `batch_invariant_ops` wheel.
+VLLM_BATCH_INVARIANT=1 pip install \
+    --extra-index-url https://mirrors.huaweicloud.com/ascend/repos/pypi/variant \
+    vllm-ascend=={{ pip_vllm_ascend_version }}
+```
+
+## Option 2: vllm-ascend is ALREADY installed
```
Comment (Ronald1995) [inline docs/source/user_guide/feature_guide/batch_invariance.md]: we need to declare the behavior of docker image

**Review by Ronald1995 — APPROVED** [review]

File: docs/source/user_guide/feature_guide/batch_invariance.md (hunk 2 above)
Comment (wangx700) [reply docs/source/user_guide/feature_guide/batch_invariance.md]: The variable is effective for both pip install . and pip install -e .: both source-build paths run build_ext, which invokes build_aclnn when COMPILE_CUSTOM_KERNELS=1; build_aclnn.sh then checks VLLM_BATCH_INVARIANT=1 and installs the batch-invariant operator packages. 
It is not effective when pip installs an already-built wheel. The documentation has been updated to distinguish these cases and show both supported source-install commands.

File: docs/source/user_guide/feature_guide/batch_invariance.md (hunk 4 above)
Comment (wangx700) [reply docs/source/user_guide/feature_guide/batch_invariance.md]: the behavior of docker image was declared in the md

File: docs/source/user_guide/feature_guide/batch_invariance.md (hunk 3 above)
Comment (wangx700) [reply docs/source/user_guide/feature_guide/batch_invariance.md:76]: A prebuilt vllm-ascend wheel does not include the `csrc` directory or `csrc/build_batch_invariant_ops.sh`. Therefore, the following commands cannot be run from the wheel's `site-packages` installation directory. `<vllm-ascend-source-dir>` must point to a matching vllm-ascend source checkout.

**Review by Ronald1995 — APPROVED** [review]
LGTM

**Review by linfeng-yuan — APPROVED** [review]
